### PR TITLE
 Add support for polymorphic bounds safe interface functions

### PR DESCRIPTION
**Pull request by Prabhuk**
Added new keyword _Itype_for_any which is a function specifier.

Pending:
1. Parsing logic of _For_any and _Itype_for_any seems to fail in linux
2. Spec needs to be updated

### CODE REVIEW

**Comment by Prabhuk** [comment]
This PR is ready for review.

File: include/clang/AST/ASTContext.h [hunk 1]
```diff
@@ -2486,12 +2488,16 @@ class ASTContext : public RefCountedBase<ASTContext> {
...
   /// checkedness.
+  /// Note:: Checked C - In bounds safe interface scopes, this function
+  /// returns true if T2 is a TypeVariableType and T1 is a pointer to void type
   bool isAtLeastAsCheckedAs(QualType T1, QualType T2) const;
 
   /// \brief Determine whether a pointer, array, or function type T1
   /// is the same as the other pointer, array, or function type T2 if
   /// checkedness is ignored.  Return true if does or false if the types
   /// differ in some other way than checkedness.
+  /// Note:: Checked C - In bounds safe interface scopes, this function
```
Comment (dtarditi) [inline include/clang/AST/ASTContext.h]: You don't need to prefix this with "Checked C", as this function is specific to the Checked C implementation.

File: include/clang/AST/ASTContext.h [hunk 2]
```diff
@@ -2486,12 +2488,16 @@ class ASTContext : public RefCountedBase<ASTContext> {
   /// at least as much checking as the other type T2.  Return true if it does
   /// or false if it does not or the types differ in some other way than
   /// checkedness.
+  /// Note:: Checked C - In bounds safe interface scopes, this function
```
Comment (dtarditi) [inline include/clang/AST/ASTContext.h]: You don't need to prefix this with "Checked C", as this function is specific to the Checked C implementation.

File: lib/AST/ASTContext.cpp [hunk 3]
```diff
@@ -8146,17 +8147,26 @@ QualType ASTContext::mergeFunctionTypes(QualType lhs, QualType rhs,
...
   if (lproto && rproto) { // two C99 style function prototypes
     assert(!lproto->hasExceptionSpec() && !rproto->hasExceptionSpec() &&
            "C++ shouldn't be here");
     // Compatible functions must have the same number of parameters
     if (lproto->getNumParams() != rproto->getNumParams())
       return QualType();
 
-    // Compatible functions must have the same number of type variables.
-    if (lproto->getNumTypeVars() != rproto->getNumTypeVars())
-      return QualType();
```
Comment (dtarditi) [inline lib/AST/ASTContext.cpp]: Place put a space after the 'if' keyword.

File: lib/AST/ASTContext.cpp [hunk 4]
```diff
@@ -8194,6 +8204,12 @@ QualType ASTContext::mergeFunctionTypes(QualType lhs, QualType rhs,
         } else if (!rReturnAnnots.IsEmpty() && lReturnAnnots.IsEmpty()) {
           ReturnAnnots = rReturnAnnots;
           allLTypes = false;
+        } else if (lproto->isItypeGenericFunction() && !rproto->isItypeGenericFunction()) {
```
Comment (dtarditi) [inline lib/AST/ASTContext.cpp]: I think you need to check that rproto is also not a generic function.  Generic function types should not be compatible with itype generic function type.  We will allow itype generic function types to be converted to generic types at assignments, call sites, and returns. We will not allow the reverse.

I think you should leave the code "as is" and open an issue to handle this after this PR is complete.    
This is already a large pull request.    You'll need to write some tests and alter some of the assignment compatibility code.  It's a special case that can wait.

File: lib/Parse/ParseDecl.cpp [hunk 5]
```diff
@@ -7204,6 +7220,91 @@ void Parser::ParseCheckedPointerSpecifiers(DeclSpec &DS) {
         Diag(StartLoc, DiagID) << PrevSpec;
 }
 
+void Parser::ParseItypeforanySpecifier(DeclSpec &DS) {
```
Comment (dtarditi) [inline lib/Parse/ParseDecl.cpp]: I think you should common this code with parsing for regular forall functions.   You can use formatting specifiers in clang error messages to customize parts of error message for specific cases.

File: lib/Parse/ParseExpr.cpp [hunk 6]
```diff
@@ -1432,9 +1432,10 @@ ExprResult Parser::ParseCastExpression(bool isUnaryExpression,
   // are compiling for OpenCL, we need to return an error as this implies
   // that the address of the function is being taken, which is illegal in CL.
 
-  // Check to see if resExpr is a generic function call. In this case, we must
-  // parse a list of type names that follows generic function decl identifier.
-  if (ParseGenericFunctionExpression(Res)) return ExprError();
+  // Check to see if resExpr is a itype genric or generic function call.
```
Comment (dtarditi) [inline lib/Parse/ParseExpr.cpp]: There's a typo: genric -> generic.

File: lib/Parse/Parser.cpp [hunk 7]
```diff
@@ -1204,7 +1204,8 @@ Decl *Parser::ParseFunctionDefinition(ParsingDeclarator &D,
   // Tell the actions module that we have entered a function definition with the
   // specified Declarator for the function.
   Sema::SkipBodyInfo SkipBody;
-  Scope* funcDeclScope = D.getDeclSpec().isForanySpecified() ? getCurScope()->getParent() : getCurScope();
+  Scope* funcDeclScope = (D.getDeclSpec().isForanySpecified() || D.getDeclSpec().isItypeforanySpecified())?
```
Comment (dtarditi) [inline lib/Parse/Parser.cpp]: Need a space before the '?'

File: lib/Sema/DeclSpec.cpp [hunk 8]
```diff
@@ -975,7 +975,7 @@ bool DeclSpec::setFunctionSpecUnchecked(SourceLocation Loc,
 bool DeclSpec::setFunctionSpecForany(SourceLocation Loc,
                                         const char *&PrevSpec,
                                         unsigned &DiagID) {
-  if (FS_forany_specified) {
+  if (FS_itypeforany_specified || FS_forany_specified) {
```
Comment (dtarditi) [inline lib/Sema/DeclSpec.cpp]: Is the case for itypeforany_specified a conflicting declspec, not a duplicate declspec?

File: lib/Sema/DeclSpec.cpp [hunk 9]
```diff
@@ -985,6 +985,19 @@ bool DeclSpec::setFunctionSpecForany(SourceLocation Loc,
   return false;
 }
 
+bool DeclSpec::setFunctionSpecItypeforany(SourceLocation Loc,
+                                              const char *&PrevSpec,
+                                              unsigned &DiagID) {
+  if (FS_itypeforany_specified || FS_forany_specified) {
```
Comment (dtarditi) [inline lib/Sema/DeclSpec.cpp]: Same here, in terms of this being a conflict, not a duplicate.

File: lib/Sema/SemaDecl.cpp [hunk 10]
```diff
@@ -8729,6 +8737,17 @@ Sema::ActOnFunctionDeclarator(Scope *S, Declarator &D, DeclContext *DC,
     }
   }
 
+  if (D.getDeclSpec().isItypeforanySpecified()) {
```
Comment (dtarditi) [inline lib/Sema/SemaDecl.cpp]: Can you common this code with the existing case with IsForanySpecified?

File: lib/Serialization/ASTWriterStmt.cpp [hunk 11]
```diff
@@ -402,9 +402,22 @@ void ASTStmtWriter::VisitDeclRefExpr(DeclRefExpr *E) {
   Record.push_back(E->hasTemplateKWAndArgsInfo());
   Record.push_back(E->hadMultipleCandidates());
   Record.push_back(E->refersToEnclosingVariableOrCapture());
-  bool isGenericFunction = E->GetTypeArgumentInfo() != nullptr;
+  bool isGenericFunction = false;
+  if(E->getDecl() && isa<FunctionDecl>(E->getDecl()))
```
Comment (dtarditi) [inline lib/Serialization/ASTWriterStmt.cpp]: I think the brace should be on the prior line.

File: lib/AST/ASTContext.cpp [hunk 12]
```diff
@@ -8923,6 +8951,18 @@ bool ASTContext::isEqualIgnoringChecked(QualType T1, QualType T2) const {
     return true;
 
   Type::TypeClass T1TypeClass = T1->getTypeClass();
+  if(T1TypeClass == Type::Pointer)
+  {
```
Comment (dtarditi) [inline lib/AST/ASTContext.cpp]: I think the brace should be on the prior line.

File: include/clang/Basic/DiagnosticParseKinds.td [hunk 13]
```diff
@@ -1204,12 +1204,27 @@ def err_forany_type_variable_identifier_expected : Error<
 def err_forany_unexpected_token : Error<
   "unexpected token in _For_any specifier">;
 
+def err_itypeforany_comma_or_parenthesis_expected : Error<
```
Comment (dtarditi) [inline include/clang/Basic/DiagnosticParseKinds.td]: Clang development guidelines favor sharing diagnostic messages when possible, to minimize the number of different messages to manage.   Diagnostic message can contain formatting specifiers that allow the message to be customized for a specific circumstance.  We should try to follow those guidelines.

**Comment by Prabhuk** [comment]
The PR comments are addressed. 
Follow-up tasks: 
1. Create a new issue for itype_for_any and for_any function type compatibility.
2. Change the signature of malloc's signature in stdlib_checked header file and fix the uses in checked scopes in tests.

File: include/clang/AST/ASTContext.h [hunk 14]
```diff
@@ -2486,12 +2488,16 @@ class ASTContext : public RefCountedBase<ASTContext> {
   /// at least as much checking as the other type T2.  Return true if it does
   /// or false if it does not or the types differ in some other way than
   /// checkedness.
+  /// Note:: In bounds safe interface scopes, this function
```
Comment (dtarditi) [inline include/clang/AST/ASTContext.h]: T1 and T2 in this comment should be reversed.

**Pull request merged**